# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit/MacroAssembler.cpp
# Commit: 296b7f5a0713
# Full Hash: 296b7f5a07133f9cbbd28ab20c1468800b0e1d39
# Author: Dmitry Bezhetskov <dbezhetskov@igalia.com>
# Date: 2020-08-06 15:44:30
# Regressor Bug: 1639153
# File Overlap Count: 1
# Description:
#   Bug 1639153 - Part 2: Preserve callee and caller tls'es. r=lth
#   
#   This is a followup patch for removing Frame::tls.
#   Now, we are preserving caller's and callee's tls'es for all possible cross-instance calls in the previously allocated abi slots.
#   We also use preserved tls values to restore the caller's tls in Ion. Baseline doesn't need this because it restores the caller tls from its private stack slot after the call.
# ==============================================================================

diff -r f68854b70d6a -r 296b7f5a0713 js/src/jit/MacroAssembler.cpp
--- a/js/src/jit/MacroAssembler.cpp	Thu Aug 06 04:57:48 2020 +0000
+++ b/js/src/jit/MacroAssembler.cpp	Thu Aug 06 05:23:07 2020 +0000
@@ -3520,6 +3520,9 @@
 
 CodeOffset MacroAssembler::wasmCallImport(const wasm::CallSiteDesc& desc,
                                           const wasm::CalleeDesc& callee) {
+  storePtr(WasmTlsReg, Address(getStackPointer(),
+                               wasm::FrameWithTls::callerTLSABIOffset()));
+
   // Load the callee, before the caller's registers are clobbered.
   uint32_t globalDataOffset = callee.importGlobalDataOffset();
   loadWasmGlobalPtr(globalDataOffset + offsetof(wasm::FuncImportTls, code),
@@ -3538,6 +3541,8 @@
   // Switch to the callee's TLS and pinned registers and make the call.
   loadWasmGlobalPtr(globalDataOffset + offsetof(wasm::FuncImportTls, tls),
                     WasmTlsReg);
+  storePtr(WasmTlsReg, Address(getStackPointer(),
+                               wasm::FrameWithTls::calleeTLSABIOffset()));
   loadWasmPinnedRegsFromTls();
 
   return call(desc, ABINonArgReg0);
@@ -3548,6 +3553,10 @@
     wasm::SymbolicAddress builtin, wasm::FailureMode failureMode) {
   MOZ_ASSERT(instanceArg != ABIArg());
 
+  storePtr(WasmTlsReg, Address(getStackPointer(),
+                               wasm::FrameWithTls::callerTLSABIOffset()));
+  storePtr(WasmTlsReg, Address(getStackPointer(),
+                               wasm::FrameWithTls::calleeTLSABIOffset()));
   if (instanceArg.kind() == ABIArg::GPR) {
     loadPtr(Address(WasmTlsReg, offsetof(wasm::TlsData, instance)),
             instanceArg.gpr());
@@ -3654,7 +3663,11 @@
     addPtr(index, scratch);
   }
 
+  storePtr(WasmTlsReg, Address(getStackPointer(),
+                               wasm::FrameWithTls::callerTLSABIOffset()));
   loadPtr(Address(scratch, offsetof(wasm::FunctionTableElem, tls)), WasmTlsReg);
+  storePtr(WasmTlsReg, Address(getStackPointer(),
+                               wasm::FrameWithTls::calleeTLSABIOffset()));
 
   Label nonNull;
   branchTest32(Assembler::NonZero, WasmTlsReg, WasmTlsReg, &nonNull);